contracts/modules/andromeda-curve/src/contract.rs

```diff
@@ -0,0 +1,280 @@
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    let resp = ADOContract::default().instantiate(
+        deps.storage,
+        env,
+        deps.api,
+        &deps.querier,
+        info,
+        BaseInstantiateMsg {
+            ado_type: CONTRACT_NAME.to_string(),
+            ado_version: CONTRACT_VERSION.to_string(),
+            kernel_address: msg.kernel_address,
+            owner: msg.owner,
+        },
+    )?;
+    RESTRICTION.save(deps.storage, &msg.restriction)?;
+    CURVE_TYPE.save(deps.storage, &msg.curve_type)?;
+
+    IS_CONFIGURED_EXP.save(deps.storage, &false)?;
+
+    Ok(resp)
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    let ctx = ExecuteContext::new(deps, info, env);
+    match msg {
+        ExecuteMsg::AMPReceive(pkt) => {
+            ADOContract::default().execute_amp_receive(ctx, pkt, handle_execute)
+        }
+        _ => handle_execute(ctx, msg),
+    }
+}
+
+fn handle_execute(mut ctx: ExecuteContext, msg: ExecuteMsg) -> Result<Response, ContractError> {
+    let action = msg.as_ref().to_string();
+
+    let action_response = call_action(
+        &mut ctx.deps,
+        &ctx.info,
+        &ctx.env,
+        &ctx.amp_ctx,
+        msg.as_ref(),
+    )?;
+
+    let res = match msg.clone() {
+        ExecuteMsg::UpdateCurveType { curve_type } => {
+            execute_update_curve_type(ctx, curve_type, action)
+        }
+        ExecuteMsg::UpdateRestriction { restriction } => {
+            execute_update_restriction(ctx, restriction, action)
+        }
+        ExecuteMsg::ConfigureExponential {
+            curve_id,
+            base_value,
+            multiple_variable_value,
+            constant_value,
+        } => execute_configure_exponential(
+            ctx,
+            curve_id,
+            base_value,
+            multiple_variable_value,
+            constant_value,
+            action,
+        ),
+        ExecuteMsg::Reset {} => execute_reset(ctx, action),
+        _ => ADOContract::default().execute(ctx, msg),
+    }?;
+
+    Ok(res
+        .add_submessages(action_response.messages)
+        .add_attributes(action_response.attributes)
+        .add_events(action_response.events))
+}
+
+pub fn execute_update_curve_type(
+    ctx: ExecuteContext,
+    curve_type: CurveType,
+    action: String,
+) -> Result<Response, ContractError> {
+    let sender = ctx.info.sender.clone();
+    ensure!(
+        has_permission(ctx.deps.storage, &sender)?,
+        ContractError::Unauthorized {}
+    );
+    CURVE_TYPE.save(ctx.deps.storage, &curve_type)?;
+
+    CURVE_ID.remove(ctx.deps.storage);
```

Comment:
This looks a little unusual. Is there no way to store this as an interface rather than individual state values?

Reply:
I resolved it. Thanks for your review